File: web/src/lib/rules.ts
// 광고 운영 룰 엔진. backend/rules.py 포팅 (순수 로직).
export type Tier = { roasMin: number; roasMax: number | null; changePct: number; label: string };
export type BidParams = {
  targetRoas?: number; minClicks?: number; maxChangePct?: number; bidFloor?: number;
  tiers?: Tier[];
};
export type KeywordParams = BidParams & {
  minImpressions?: number; lowImpBoost?: number;
  lowRankCtrThreshold?: number; lowRankThreshold?: number; lowRankBoost?: number;
};

function roundTo10(v: number, floor: number): number {
  return Math.max(floor, Math.floor(v / 10) * 10);
}

function capChange(current: number, recommended: number, maxChangePct: number): number {
  const up = Math.floor(current * (1 + maxChangePct / 100));
  const down = Math.floor(current * (1 - maxChangePct / 100));
  return Math.max(Math.min(recommended, up), down);
}

function matchTier(tiers: Tier[], roas: number): Tier | null {
  for (const t of tiers) {
    const lo = t.roasMin ?? 0;
    const hi = t.roasMax;
    if (roas >= lo && (hi === null || hi === undefined || roas < hi)) return t;
  }
  return null;
}

export function recommendBid(args: {
  currentBid: number; roas7: number; clicks7: number; params: BidParams;
}): { bid: number | null; reason: string } {
  const { currentBid, roas7, clicks7, params } = args;
  const minClicks = params.minClicks ?? 10;
  const maxChange = params.maxChangePct ?? 30;
  const floor = params.bidFloor ?? 70;
  const tiers = params.tiers ?? [];

  if (clicks7 < minClicks) return { bid: null, reason: `클릭 ${clicks7}회 < ${minClicks}회 (데이터 부족)` };

  const tier = matchTier(tiers, roas7);
  if (!tier) return { bid: null, reason: `ROAS ${roas7.toFixed(0)}% · 매칭 티어 없음` };

  const changePct = tier.changePct ?? 0;
  const label = tier.label ?? `${changePct >= 0 ? "+" : ""}${changePct}%`;
  if (changePct === 0) return { bid: null, reason: `ROAS ${roas7.toFixed(0)}% · ${label}` };

  const target = capChange(currentBid, Math.floor(currentBid * (1 + changePct / 100)), maxChange);
  return { bid: roundTo10(target, floor), reason: `ROAS ${roas7.toFixed(0)}% · ${label}` };
}

export function recommendKeywordBid(args: {
  currentBid: number; imp30: number; clk30: number; sales30: number;
  conv30: number; avgRnk30: number; params: KeywordParams;
}): { bid: number | null; reason: string } {
  const { currentBid, imp30, clk30, sales30, conv30, avgRnk30, params } = args;
  const minClicks = params.minClicks ?? 5;
  const minImp = params.minImpressions ?? 100;
  const maxChange = params.maxChangePct ?? 30;
  const floor = params.bidFloor ?? 70;
  const lowImpBoost = params.lowImpBoost ?? 20;
  const lowRankCtrThr = params.lowRankCtrThreshold ?? 1.0;
  const lowRankThr = params.lowRankThreshold ?? 10.0;
  const lowRankBoost = params.lowRankBoost ?? 20;
  const tiers = params.tiers ?? [];

  const ctr = imp30 ? (clk30 / imp30) * 100 : 0;
  const roas = sales30 ? (conv30 / sales30) * 100 : 0;

  if (imp30 === 0) return { bid: null, reason: "노출 0회 (신규/집행 직전)" };

  if (imp30 < minImp) {
    const target = capChange(currentBid, Math.floor(currentBid * (1 + lowImpBoost / 100)), maxChange);
    return { bid: roundTo10(target, floor), reason: `노출 ${imp30}회 < ${minImp} · +${lowImpBoost}% (테스트 확대)` };
  }

  if (avgRnk30 > lowRankThr && ctr < lowRankCtrThr) {
    const target = capChange(currentBid, Math.floor(currentBid * (1 + lowRankBoost / 100)), maxChange);
    return { bid: roundTo10(target, floor), reason: `평균순위 ${avgRnk30.toFixed(1)}위 · CTR ${ctr.toFixed(1)}% · +${lowRankBoost}% (순위 부족)` };
  }

  if (clk30 < minClicks) return { bid: null, reason: `클릭 ${clk30}회 < ${minClicks} (CTR ${ctr.toFixed(1)}%, 수집 중)` };

  const tier = matchTier(tiers, roas);
  if (!tier) return { bid: null, reason: `ROAS ${roas.toFixed(0)}% · 매칭 티어 없음` };
  const changePct = tier.changePct ?? 0;
  const label = tier.label ?? `${changePct >= 0 ? "+" : ""}${changePct}%`;
  if (changePct === 0) return { bid: null, reason: `ROAS ${roas.toFixed(0)}% · ${label}` };
  const target = capChange(currentBid, Math.floor(currentBid * (1 + changePct / 100)), maxChange);
  return { bid: roundTo10(target, floor), reason: `ROAS ${roas.toFixed(0)}% · ${label}` };
}

// ---------- rules_store 해결 로직 ----------
export type RuleScope =
  | { type: "global" }
  | { type: "media"; campaignTps: string[] }
  | { type: "campaign"; campaignIds: string[] };
export type Rule<P> = { id: string; name: string; enabled: boolean; scope: RuleScope; params: P };
export type RulesStore = { bid: Rule<BidParams>[]; keyword: Rule<KeywordParams>[] };

export function resolveRule<P>(
  rules: Rule<P>[], campaignId: string, campaignTp: string,
): Rule<P> | null {
  const active = rules.filter((r) => r.enabled);
  const camp: Rule<P>[] = [];
  const media: Rule<P>[] = [];
  const global: Rule<P>[] = [];
  for (const r of active) {
    const s = r.scope;
    if (s.type === "campaign" && s.campaignIds.includes(campaignId)) camp.push(r);
    else if (s.type === "media" && s.campaignTps.includes(campaignTp)) media.push(r);
    else if (s.type === "global") global.push(r);
  }
  return camp[0] ?? media[0] ?? global[0] ?? null;
}

export const DEFAULT_RULES: RulesStore = {
  bid: [{
    id: "default-bid", name: "기본 입찰가 규칙", enabled: true, scope: { type: "global" },
    params: {
      targetRoas: 250, minClicks: 10, maxChangePct: 30, bidFloor: 70,
      tiers: [
        { roasMin: 400, roasMax: null, changePct: 20, label: "ROAS 400% 이상: +20% 확대" },
        { roasMin: 250, roasMax: 400, changePct: 0, label: "목표 달성: 유지" },
        { roasMin: 100, roasMax: 250, changePct: -15, label: "목표 미달: -15%" },
        { roasMin: 0, roasMax: 100, changePct: -30, label: "적자: -30%" },
      ],
    },
  }],
  keyword: [{
    id: "default-keyword", name: "기본 키워드 규칙", enabled: true, scope: { type: "global" },
    params: {
      targetRoas: 250, minClicks: 5, minImpressions: 100, maxChangePct: 30, bidFloor: 70,
      lowImpBoost: 20, lowRankCtrThreshold: 1.0, lowRankThreshold: 10.0, lowRankBoost: 20,
      tiers: [
        { roasMin: 400, roasMax: null, changePct: 20, label: "ROAS 400% 이상: +20% 확대" },
        { roasMin: 250, roasMax: 400, changePct: 0, label: "목표 달성: 유지" },
        { roasMin: 100, roasMax: 250, changePct: -15, label: "목표 미달: -15%" },
        { roasMin: 0, roasMax: 100, changePct: -30, label: "적자: -30%" },
      ],
    },
  }],
};

File: backend/rules.py
"""광고 운영 룰 엔진. 파라미터는 rules_store에서 주입."""


def _round_to_10(value: int, floor: int) -> int:
    return max(floor, (value // 10) * 10)


def _cap_change(current: int, recommended: int, max_change_pct: float) -> int:
    max_up = int(current * (1 + max_change_pct / 100))
    max_down = int(current * (1 - max_change_pct / 100))
    return max(min(recommended, max_up), max_down)


def _match_tier(tiers: list[dict], roas: float) -> dict | None:
    for t in tiers:
        lo = t.get("roasMin", 0)
        hi = t.get("roasMax")
        if roas >= lo and (hi is None or roas < hi):
            return t
    return None


def recommend_bid(*, current_bid: int, roas7: float, clicks7: int, params: dict) -> tuple[int | None, str]:
    """광고그룹 입찰가 추천. params는 rules_store의 bid 규칙 params."""
    min_clicks = params.get("minClicks", 10)
    max_change = params.get("maxChangePct", 30)
    floor = params.get("bidFloor", 70)
    tiers = params.get("tiers", [])

    if clicks7 < min_clicks:
        return None, f"클릭 {clicks7}회 < {min_clicks}회 (데이터 부족)"

    tier = _match_tier(tiers, roas7)
    if not tier:
        return None, f"ROAS {roas7:.0f}% · 매칭 티어 없음"

    change_pct = tier.get("changePct", 0)
    label = tier.get("label", f"{change_pct:+d}%")
    if change_pct == 0:
        return None, f"ROAS {roas7:.0f}% · {label}"

    target = int(current_bid * (1 + change_pct / 100))
    target = _cap_change(current_bid, target, max_change)
    return _round_to_10(target, floor), f"ROAS {roas7:.0f}% · {label}"


def recommend_keyword_bid(*, current_bid: int, imp30: int, clk30: int, sales30: int,
                          conv30: int, avgRnk30: float, params: dict) -> tuple[int | None, str]:
    """키워드 입찰가 추천 (30일 기준). params는 rules_store의 keyword 규칙 params."""
    min_clicks = params.get("minClicks", 5)
    min_imp = params.get("minImpressions", 100)
    max_change = params.get("maxChangePct", 30)
    floor = params.get("bidFloor", 70)
    low_imp_boost = params.get("lowImpBoost", 20)
    low_rank_ctr_thr = params.get("lowRankCtrThreshold", 1.0)
    low_rank_thr = params.get("lowRankThreshold", 10.0)
    low_rank_boost = params.get("lowRankBoost", 20)
    tiers = params.get("tiers", [])

    ctr = (clk30 / imp30 * 100) if imp30 else 0.0
    roas = (conv30 / sales30 * 100) if sales30 else 0.0

    if imp30 == 0:
        return None, "노출 0회 (신규/집행 직전)"

    if imp30 < min_imp:
        target = _cap_change(current_bid, int(current_bid * (1 + low_imp_boost / 100)), max_change)
        return _round_to_10(target, floor), f"노출 {imp30}회 < {min_imp} · +{low_imp_boost}% (테스트 확대)"

    if avgRnk30 > low_rank_thr and ctr < low_rank_ctr_thr:
        target = _cap_change(current_bid, int(current_bid * (1 + low_rank_boost / 100)), max_change)
        return _round_to_10(target, floor), f"평균순위 {avgRnk30:.1f}위 · CTR {ctr:.1f}% · +{low_rank_boost}% (순위 부족)"

    if clk30 < min_clicks:
        return None, f"클릭 {clk30}회 < {min_clicks} (CTR {ctr:.1f}%, 수집 중)"

    tier = _match_tier(tiers, roas)
    if not tier:
        return None, f"ROAS {roas:.0f}% · 매칭 티어 없음"
    change_pct = tier.get("changePct", 0)
    label = tier.get("label", f"{change_pct:+d}%")
    if change_pct == 0:
        return None, f"ROAS {roas:.0f}% · {label}"
    target = _cap_change(current_bid, int(current_bid * (1 + change_pct / 100)), max_change)
    return _round_to_10(target, floor), f"ROAS {roas:.0f}% · {label}"
